woven-planet/wicker
# Abstractify the Path Factory

## Overview

Abstractifies the file pathing class in Wicker. Wicker has one common pathing structure and it doesn't constrain itself to S3. This abstraction allows for adding different file pathers on top of this if needed while keeping the consistent structure.

## Motivation

This set of PRs, https://github.com/woven-planet/wicker/pull/56, https://github.com/woven-planet/wicker/pull/57 aim to fix two problems/incorrect assumptions. Below is the context on the problems bulleted:

1. Wicker right now limits itself very heavily to S3 usage. It only implements S3 and more or less lacks the infrastructure/correct abstractions to add more data sources. This is evidenced by classes like [`S3Dataset`](https://github.com/woven-planet/wicker/blob/18dcee4a8691ded85925219861219443d355028c/wicker/core/datasets.py#L49) and [`S3PathFactory`](https://github.com/woven-planet/wicker/blob/18dcee4a8691ded85925219861219443d355028c/wicker/core/storage.py#L173) which can be seen as implementations of a base class but lack the base class. These are the only writing and access methods and there is no common class for interfacing.
2. There is no common core, there is no common API that users can infer from the different classes. It's very easy with each implementation to diverge from the common access pattern. This should be rectified and a common core added.
3. Wicker has a very tight file structure, it shouldn't deviate from data platform to data platform. What should deviate is the location in which that file structure is stored and the way wicker accesses it but not the actual path structure.
4. There is no way to have a local/mounted file structure. Currently Wicker doesn't even support having a file system locally with a dataset. You can't create a dataset locally and test it, that's a huge problem for functional testing that we have to rely on mocking out S3 just to test locally.

## What this implements

1. This PR is very targeted to abstracting away the `S3PathFactory` into an implementation class for S3 and a base class `WickerPathFactory`. This new path defines the path structure generically while the `S3PathFactory` only appends on the relevant S3 pieces. This gives us a common class to build off where we can keep the access pattern for paths identical across both for easy swapping of user code.

## Testing

- Run CI
- Run your workflow
- Verify outputs

## Compatibility

This PR is entirely backward compatible because it does not change any access patterns. The function signatures, names, and outputs are 1-1 so there is no requirement to change anything for user code if the upgraded is undertaken.
